### DISCUSSION TITLE
tasks: enforcing A happens after B, when A doesn't package-depend on B

### DISCUSSION THREAD

**Discussion by fauxbytes** [vercel/turborepo#9433] [Help] [answered]
### Summary

Hi. 
given the following repo, containg a set of apps, and a separate 'oper' workspace:
```
repo/
├── apps
│   ├── func-a
│   │   └── package.json
│   ├── func-b
│   │   └── package.json
│   └── func-c
│       └── package.json
├── oper
│   └── package.json
├── package.json
└── turbo.json
```

Each app has a build script, which can all be triggered via a standard `build` task in `turbo.json`:
```
    "build": {
      "dependsOn": ["^build"],
      "outputs": ["dist/**"]
    }
```

The `oper` project deploys the apps to a cloud service. Like the apps, it has a `build` step, but that step is not dependent on the apps' `build`. `oper` also has no code dependency on any of the apps, so there's no `dependencies` relationship declared between them in `oper/package.json`. Now, oper also has a `deploy` script. This script takes the build products of the apps and publishes them to the cloud service. This script _is_ dependent on all the apps having been built. I.E "_run oper:deploy only after apps/*:build have finished successfully_". Unsure how/if to declare that relationship in turbo. ATM it isn't. My workaround is a script on the root `package.json`, which calls the two (independent) tasks sequentially:
```
"scripts": {
  "deploy": "turbo build && turbo deploy"
}
```

turbo.json:
```
"tasks": {
    "deploy": {
      "cache": false
    }
}
```

I'm not very satisfied with this solution, as it hardly utilizes turbo's task management for deploying (I could essentially drop `turbo deploy` and just call oper:deploy using the package manager).

Would appreciate suggestions for improvement.

**Comment by anthonyshew** [comment]
Would [depending on a specific task in a specific package](https://turbo.build/repo/docs/crafting-your-repository/configuring-tasks#depending-on-a-specific-task-in-a-specific-package) help in this case? Sorry if that's not the solution. Trying to wrap my head around what you're describing. 😄

**Comment by fauxbytes** [comment]
@anthonyshew thanks. Looked at it, but could not get it to solve the above. Cannot single-out a specific task that `oper` could `dependOn` that would deliver the required sequence. But if I'm not mistaken, the key to achieving such a sequence is to link packages that are unrelated through dependency.

**Reply by anthonyshew** [reply]
Reading this for the second time, I think I'm understanding it better. Its ounds like:

```
{
  "tasks": {
      "build": {
        "dependsOn": ["^build"]
      }
    },
    {
      "oper#build": {
         "dependsOn": []
    },
    {
      "oper#deploy: {
        "dependsOn": ["^build"],
    }
}
```

You'd then install the applications into the `oper` package, so that `^build` can keep knowledge of those dependencies.

The key ingredient is that the Task Graph is built off of the Package Graph. So, while there aren't any _code_ dependencies, there are _task_ dependencies, and installing those packages into `oper` lets that relationship be expressed.

Note: Where I'm doing `oper#`, you could instead use [Package Configurations](https://turbo.build/repo/docs/reference/package-configurations) to localize those task definitions.

**Comment by fauxbytes** [comment, marked as answer]
> install the applications into the oper package, so that ^build can keep knowledge of those dependencies

Sorry. Unsure what is meant by that.

[Attached](https://github.com/user-attachments/files/17769491/ar.zip) is a sample repo with the suggested `turbo.json`. 

build works as expected:
```
$ turbo run build --concurrency=1 --output-logs=hash-only
turbo 2.3.0

• Packages in scope: @repo/shared, app-1, app-2, app-3, oper
• Running build in 5 packages
• Remote caching disabled
@repo/shared:build: cache miss, executing 805e425a363c1918
oper:build: cache miss, executing 30e4db4b9f4aac2a
app-3:build: cache miss, executing 9e1522c3356554db
app-1:build: cache miss, executing 44eabcb451aa6b56
app-2:build: cache miss, executing 6a3dc65a9437d342

 Tasks:    5 successful, 5 total
Cached:    0 cached, 5 total
  Time:    6.532s
```
deploy seems like it doesn't:
```
$ turbo run deploy --concurrency=1 --force
turbo 2.3.0

• Packages in scope: @repo/shared, app-1, app-2, app-3, oper
• Running deploy in 5 packages
• Remote caching disabled

No tasks were executed as part of this run.

 Tasks:    0 successful, 0 total
Cached:    0 cached, 0 total
  Time:    489ms
```

**Reply by anthonyshew** [reply]
If you edit `oper/package.json` to this, it appears to describe the behavior you're looking for:
```
{
  "name": "oper",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "build": "echo oper:build",
    "deploy": "echo oper:deploy"
  },
  "dependencies": {
    "app-1": "workspace:*",
    "app-2": "workspace:*",
    "app-3": "workspace:*"
  },
  "keywords": [],
  "author": "",
  "license": "ISC"
}

```
